stripe/src/main/java/com/stripe/android/IssuingCardPinService.java

```diff
@@ -0,0 +1,268 @@
+package com.stripe.android;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import com.stripe.android.exception.APIConnectionException;
+import com.stripe.android.exception.APIException;
+import com.stripe.android.exception.AuthenticationException;
+import com.stripe.android.exception.CardException;
+import com.stripe.android.exception.InvalidRequestException;
+
+import org.json.JSONException;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/*
+ * Methods for retrieval / update of a Stripe Issuing card
+ */
+public class IssuingCardPinService
+        implements EphemeralKeyManager.KeyManagerListener<IssuingCardEphemeralKey> {
+
+    private static final long KEY_REFRESH_BUFFER_IN_SECONDS = 30L;
+    private static final String PIN_RETRIEVE = "PIN_RETRIEVE";
+    private static final String PIN_UPDATE = "PIN_UPDATE";
+
+    private static IssuingCardPinService mInstance;
+    private @NonNull
+    EphemeralKeyManager<IssuingCardEphemeralKey> mEphemeralKeyManager;
+    private IssuingCardPinRetrievalListener mCardPinRetrievalListener;
+    private IssuingCardPinUpdateListener mCardPinUpdateListener;
+
+    private IssuingCardPinService(
+            @NonNull EphemeralKeyProvider keyProvider
+    ) {
+        mEphemeralKeyManager = new EphemeralKeyManager<>(
+                keyProvider,
+                this,
+                KEY_REFRESH_BUFFER_IN_SECONDS,
+                null,
+                IssuingCardEphemeralKey.class);
+    }
+
+    /**
+     * Create a IssuingCardPinService with the provided {@link EphemeralKeyProvider}.
+     *
+     * @param keyProvider an {@link EphemeralKeyProvider} used to get
+     *                    {@link CustomerEphemeralKey EphemeralKeys} as needed
+     */
+    public static IssuingCardPinService init(@NonNull EphemeralKeyProvider keyProvider) {
+        mInstance = new IssuingCardPinService(keyProvider);
+        return mInstance;
+    }
+
+    /**
+     * Retrieves a PIN for a given card
+     *
+     * @param cardId          the ID of the card (looks like ic_abcdef1234)
+     * @param verificationId  the ID of the verification that was sent to the cardholder
+     *                        (typically server-side, through /v1/issuing/verifications)
+     * @param userOneTimeCode the one-time code that was sent to the cardholder through sms or email
+     * @param listener        a listener for either the PIN, or any error that can occur
+     */
+    public void retrievePin(
+            @NonNull String cardId,
+            @NonNull String verificationId,
+            @NonNull String userOneTimeCode,
+            @Nullable IssuingCardPinRetrievalListener listener
+    ) {
+        mCardPinUpdateListener = null;
+        mCardPinRetrievalListener = listener;
```

Comment:
instead of setting listeners as instance variables, can we pass them through the method calls?

something like:

`mEphemeralKeyManager.retrieveEphemeralKey(PIN_RETRIEVE, arguments, listener);` -> `onKeyUpdate(..., listener);`

My concern is that the current implementation can result in a race condition, say updatePin is called before retrievePin completes.

Reply:
Let's try that it's a good idea, there is room for a race condition, the way this API would be used makes it kinda unlikely, but your idea is very good

Reply:
OK I've tried passing the listener through the arguments. It's untyped as it goes through which makes for some gymnastic getting it back, but it should avoid race conditions.

Note that I had to modify `EphemeralKeyManager` and `CustomerSession` (in places that were not publicly exposed) to achieve this